logstash-core/src/test/java/org/logstash/FileLockFactoryTest.java

```diff
@@ -88,4 +95,50 @@ public void ReleaseUnobtainedLock() throws IOException {
         FileLockFactory.getDefault().releaseLock(lock);
         FileLockFactory.getDefault().releaseLock(lock);
     }
+
+    @Test
+    public void crossJvmObtainLockOnLocked() throws Exception {
+        Process p = null;
+        ExecutorService executor = null;
+        String lockFile = ".testCrossJvm";
+        FileLock lock = null;
+
+        // Build the command to spawn a children JVM.
+        String[] cmd = {
+            Paths.get(System.getProperty("java.home"), "bin", "java").toString(),
+            "-cp", System.getProperty("java.class.path"),
+            Class.forName("org.logstash.FileLockFactoryMain").getName(),
+            lockDir, lockFile
+        };
+
+        try {
+            // Start the children program that will lock the file.
+            p = new ProcessBuilder(cmd).start();
+            InputStream is = p.getInputStream();
+            /* Wait the children program write to stdout, meaning the file
+             * is locked. Set a timeout to ensure it returns.
+             */
+            executor = Executors.newSingleThreadExecutor();
+            Future<Integer> future = executor.submit(() -> {return is.read();});
+            assertTrue(future.get(30, TimeUnit.SECONDS) > -1);
+
+            // Check the children process is still running.
+            assertThat(p.isAlive(), is(equalTo(true)));
+
+            try {
+                // Try to obtain the lock held by the children process.
+                FileLockFactory.getDefault().obtainLock(lockDir, lockFile);
+                fail("Should have threw an exception");
+            } catch (LockException e) {
+                // Expected exception as the file is already locked.
+            }
+        } finally {
+            if (p != null) {
+                p.destroy();
+            }
+            if (executor != null) {
+                executor.shutdownNow();
```

Comment:
@wainersm I think we should use this (https://github.com/elastic/logstash/blob/master/logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java#L47) patter here too.

The simple reason being that `shutdownNow()` doesn't block and I think it's generally kind of dirty to throw in a `finally` block (makes things really confusing, in this case it would mean that we have to be on the lookout that `p.destroy` never throws for example). I think destroying executors in a safe way in `@After` is cleaner here and 100% prevents threads leaking from one thread to the next (I recommend running the Apache Spark testsuit locally if you want to observe an example of how nasty this can become :D ).